### PR TITLE
Updated unit tests

### PR DESCRIPTION
**Pull request by ryanbarker**
During work on my previous ticket, I realised that some of the unit tests weren't testing the intended test case. In `test_controls.py` and `test_pulses.py` there are tests that check data validation by sending invalid data and expecting a `DocumentError`. While the exception was being raised, it was being raised for different reasons, e.g. checking if the `upper_bound` is too large, but the request was failing due to `segment_type` missing, etc.

I've added `JsonApiErrorResponseChecker` which can check the response to confirm the request is failing for the expected reason.

I've also merged test functions with common implementation using `pytest.mark.parametrize` so it will be easier to add more cases in future and used `qctrl` and `system` as fixtures instead of importing them into each test file.

### CODE REVIEW

File: tests/utils.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+        assert isinstance(data["errors"], list), data
+        assert len(data["errors"]) == 1, data
+        error = data["errors"][0]
+
+        # check error detail
+        if self.expected_detail:
+            try:
+                actual_detail = error["detail"]
+            except KeyError:
+                raise ValueError("error missing [detail]: {}".format(error))
```
Comment (stabacco) [inline tests/utils.py]: What's the reasoning behind raising a different exception here?

**Comment by luigibertaco** [comment]
@ryanbarker I like the way you created the JsonApiErrorResponseChecker, however, I think I will not work with the newest python-api-client.

The new client will return the error wrapped with QctrlApiException(QctrlExpeption)

You can still access the original issue using from __cause__

File: tests/test_control_noises.py [hunk 2]
```diff
@@ -13,9 +13,10 @@
 # limitations under the License.
 import numpy as np
 import pytest
-from utils import (
-    qctrl,
-    system, 
+
+from jsonapi_client.exceptions import DocumentError
```
Comment (stabacco) [inline tests/test_control_noises.py]: shouldn't these be wrapped up with @luigibertaco 's exceptions now?

File: tests/conftest.py [hunk 3]
```diff
@@ -11,6 +11,20 @@
...
-sys.path.append(os.path.join(os.path.dirname(__file__), 'helpers'))
+import pytest
+
+from qctrl import Qctrl
+
+
+@pytest.fixture(scope="session")
+def qctrl():
+    key = os.environ.get("QCTRL_TEST_TOKEN")
+    api_host = os.environ.get("QCTRL_API_HOST")
```
Comment (stabacco) [inline tests/conftest.py]: maybe we could double check that these variables exist and raise an EnvironmentError if missing.

File: tests/utils.py (hunk 1 above)
Comment (ryanbarker) [reply tests/utils.py]: Just wanted to also give output of the entire error dict. It probably should still be a KeyError I guess.

**Comment by ryanbarker** [comment]
> @ryanbarker I like the way you created the JsonApiErrorResponseChecker, however, I think I will not work with the newest python-api-client.
> 
> The new client will return the error wrapped with QctrlApiException(QctrlExpeption)
> 
> You can still access the original issue from `__cause__ `

Ah ok, I'll update the qctrlapiclient version to get those changes.

File: tests/conftest.py (hunk 3 above)
Comment (ryanbarker) [reply tests/conftest.py]: good idea. I'll let api host be None though since its not required for the Qctrl object.

**Pull request merged**